starter-code/app.js

```diff
@@ -1,86 +1,106 @@
+mongoose.connect('mongodb://localhost/passport-local')
+
+// require the user model
+const User = require('./models/user')
+
+const session = require('express-session')
+const bcrypt = require('bcrypt')
+const passport = require('passport')
+const LocalStrategy = require('passport-local').Strategy
+
+const flash = require('connect-flash')
+
+const authRoutes = require('./routes/passportRouter')
+app.use('/', authRoutes)
+
+// enable sessions here
+
+app.use(session({
+  secret: 'our-passport-local-strategy-app',
+  resave: true,
+  saveUninitialized: true
+}))
+
+passport.serializeUser((user, cb) => {
+  cb(null, user._id)
+})
+
+passport.deserializeUser((id, cb) => {
+  User.findOne({ '_id': id }, (err, user) => {
+    if (err) { return cb(err) }
+    cb(null, user)
+  })
+})
+
+app.use(flash());
+passport.use(new LocalStrategy({
+  passReqToCallback: true;
+} ,(username, password, next) => {
+  User.findOne({ username }, (err, user) => {
+    if (err) {
+      return next(err)
+    }
+    if (!user) {
+      return next(null, false, { message: 'Incorrect username' })
+    }
+    if (!bcrypt.compareSync(password, user.password)) {
+      return next(null, false, { message: 'Incorrect password' })
+    }
+
+    return next(null, user)
+  })
+}))
+
+// initialize passport and session here
+app.use(passport.initialize())
+app.use(passport.session())
 
 // view engine setup
-app.set('views', path.join(__dirname, 'views'));
-app.set('view engine', 'ejs');
-
-
-app.use(logger('dev'));
-app.use(bodyParser.json());
-app.use(bodyParser.urlencoded({ extended: false }));
-app.use(cookieParser());
-app.use(express.static(path.join(__dirname, 'public')));
+app.set('views', path.join(__dirname, 'views'))
+app.set('view engine', 'ejs')
+
+app.use(logger('dev'))
+app.use(bodyParser.json())
+app.use(bodyParser.urlencoded({ extended: false }))
+app.use(cookieParser())
+app.use(express.static(path.join(__dirname, 'public')))
 // require in the routers
-app.use('/', index);
-app.use('/', users);
-app.use('/', passportRouter);
-
-
-
-
-
-//passport code here
-
-
-
-
-
-
-
-
+app.use('/', index)
+app.use('/', users)
+app.use('/', passportRouter)
 
+// passport code here
 
 // catch 404 and forward to error handler
-app.use(function(req, res, next) {
-  var err = new Error('Not Found');
-  err.status = 404;
-  next(err);
-});
+app.use(function (req, res, next) {
+  var err = new Error('Not Found')
+  err.status = 404
+  next(err)
+})
 
 // error handler
-app.use(function(err, req, res, next) {
+app.use(function (err, req, res, next) {
   // set locals, only providing error in development
-  res.locals.message = err.message;
-  res.locals.error = req.app.get('env') === 'development' ? err : {};
+  res.locals.message = err.message
+  res.locals.error = req.app.get('env') === 'development' ? err : {}
 
   // render the error page
-  res.status(err.status || 500);
-  res.render('error');
-});
+  res.status(err.status || 500)
+  res.render('error')
+})
```

Comment:
Change the error handle for the one we have in the cheatsheet:
https://github.com/ByronGBP/ih-cheat-sheet/blob/master/snippets/express-apps/app.js